### ISSUE TITLE
What is now the alternative of `useContainer` ?

### ISSUE THREAD

**Issue by Ollie1700** [typeorm/typeorm#9096]
## Documentation Issue

According to the [0.3.0 changelog](https://github.com/typeorm/typeorm/blob/master/CHANGELOG.md):

> all container-related features (UseContainerOptions, ContainedType, ContainerInterface, defaultContainer, useContainer, getFromContainer) are deprecated.

What is the alternative now? After upgrading, our code no longer works with the following error:

```
Cannot get Connection with name \"default\" from the ConnectionManager. Make sure you have created the connection and called \"useContainer(Container)\" in your application before establishing a connection and importing any entity into TypeORM.
```

Example usage that now no longer works:

Resolver Class:
```typescript
import { EntityManager } from 'typeorm'
import { InjectManager } from 'typeorm-typedi-extensions'
import Batch from '../entities_new/Batch'

@Resolver(of => Batch)
export default class BatchResolver implements ResolverInterface<Batch> {

  @InjectManager()
  private readonly em: EntityManager

  // Queries and mutations
  // .....
}
```

Initialisation code:
```typescript
import 'reflect-metadata'
import { DataSource, useContainer } from 'typeorm'
import { Container } from 'typeorm-typedi-extensions'

export default async function initEntitiesAndResolvers() {
  
  // TypeORM
  useContainer(Container) // This is now highlighted as deprecated
  const dataSource = new DataSource({
    ....
  })
  await dataSource.initialize()
}
```

### What was unclear or otherwise insufficient?

Is the only option here to downgrade TypeORM version? Is there a reason why container functionality has been removed and/or is there an alternative?

Thank you in advance for any advice.

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by paramsinghvc** [comment]
Same question here. Why did they deprecate useContainer without any explanation. It makes the whole flow broken.

**Comment by ericmorand** [comment]
Most probably because it is not in the scope of TypeORM to know anything about dependency injection.

**Comment by francas** [comment]
Also had a similar issue, here's our approach:

In `controller.ts` (entry point, but at the top of imports) we added this, followed by `.initialize()` below.

```
// We need to add DataSource to containers list before calling repositories, since repos reference this DataSource.
Container.set(DataSource, new DataSource(Database.getConnectionOptions()));
```

Repository file `user.repository.ts` then looks like this:
```
export const UserRepository = Container.get(DataSource)
    .getRepository(User)
    .extend({
        async doSomething(): Promise<void> {
        },
        async doSomethingElse(): Promise<void> {
        },
     });
```

And to call it in code you no longer do `Container.get(UserRepository).doSomething()` but directly `UserRepository.doSomething()`

Hope this helps.

**Comment by SecondThread** [comment]
Aight, I found it.

Basics are described here: https://typegraphql.com/docs/0.17.0/dependency-injection.html

TLDR:
```
await buildSchema({
  container: (({ context }: ResolverData<TContext>) => context.container);
};
```

Plus:
```
import { ApolloServer } from "apollo-server";
import { Container } from "typedi";

const server = new ApolloServer({
  // schema comes from `buildSchema` as always
  schema,
  // provide unique context with `requestId` for each request
  context: () => {
    // generate the requestId (it also may come from `express-request-id` or other middleware)
    const requestId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER); // uuid-like
    const container = Container.of(requestId); // get the scoped container
    const context = { requestId, container }; // create fresh context object
    container.set("context", context); // place context or other data in container
    return context;
  },
});
```

But they should really explain this as a comment imo.

**Comment by wallace-sf** [comment]
> Also had a similar issue, here's our approach:
> 
> In `controller.ts` (entry point, but at the top of imports) we added this, followed by `.initialize()` below.
> 
> ```
> // We need to add DataSource to containers list before calling repositories, since repos reference this DataSource.
> Container.set(DataSource, new DataSource(Database.getConnectionOptions()));
> ```
> 
> Repository file `user.repository.ts` then looks like this:
> 
> ```
> export const UserRepository = Container.get(DataSource)
>     .getRepository(User)
>     .extend({
>         async doSomething(): Promise<void> {
>         },
>         async doSomethingElse(): Promise<void> {
>         },
>      });
> ```
> 
> And to call it in code you no longer do `Container.get(UserRepository).doSomething()` but directly `UserRepository.doSomething()`
> 
> Hope this helps.

Can you provide a minimal example? It's not clear to me how you fix it.

**Comment by ericmorand** [comment]
Why would you need anything specific for Nest.js? Just instantiate a Datasource and pass it to whatever factory requires it.

**Comment by Meywether** [comment]
Hi @ericmorand,
I need it for this:
![image](https://github.com/typeorm/typeorm/assets/39849054/d4c3a40b-3489-4d9a-a478-8e228a37a9db)
Thanks a lot :)

**Comment by natalynyu** [comment]
Does anyone know what the alternative for `getFromContainer` is?

**Comment by sheygs** [comment]
> ```ts
> typeorm-typedi-extension
> ```

Would be glad if you could kindly provide an example usage

**Comment by North-City** [comment]
> Same question here. Why did they deprecate useContaine

 you can try import useContainer from  class-validator .it maybe works right.

**Comment by jeesus** [comment]
The irony is we can't really drop the container either since [internally TypeORM](https://github.com/typeorm/typeorm/blob/master/src/connection/ConnectionMetadataBuilder.ts#L66) still relies on the container to build the list of the Subscribers.

**Comment by alumni** [comment]
We actually plan to remove _most_ of the deprecated features in 1.0 later this year.

In general, things that are not in the scope of an ORM were deprecated in 0.3 - but there might be some exceptions and we'll have to review them in the next version.

Probably we'll have to figure this one out.

Setting the label "breaking change" so we don't forget about it.

**Comment by pkuczynski** [comment]
This has been addressed in #12180 — the deprecated container system (`useContainer`, `getFromContainer`, `ContainerInterface`, etc.) has been fully removed in v1.0.

The migration path is documented in the [v1 migration guide](https://typeorm.io/migration-v1#container-system): pass pre-built instances directly to `DataSource` options instead of relying on the container:

```typescript
new DataSource({
    subscribers: [new MySubscriber(dep1, dep2)],
    migrations: [new MyMigration(dep1)],
})
```

For NestJS users: NestJS has its own integration (`@nestjs/typeorm`) that handles DI independently of TypeORM's removed container system — no changes needed there.
